nvme_gw_server.py

```diff
@@ -19,47 +19,155 @@
+        spdk_process: Subprocess running SPDK NVMEoF target application
+    """
+
     def __init__(self, nvme_config):
 
         self.logger = nvme_config.logger
         self.nvme_config = nvme_config
+        self.persistent_config = OmapPersistentConfig(nvme_config)
+        self.spdk_process = None
+        self.server = None
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        """Cleans up SPDK and server instances."""
+
+        if self.spdk_process is not None:
+            self.logger.info("Terminating SPDK...")
+            self.spdk_process.terminate()
+            try:
+                timeout = self.nvme_config.getfloat("spdk", "timeout")
+                self.spdk_process.communicate(timeout=timeout)
+            except subprocess.TimeoutExpired:
+                self.spdk_process.kill()
+
+        if self.server is not None:
+            self.logger.info("Stopping the server...")
+            self.server.stop(None)
+
+        self.logger.info("Exiting the gateway process.")
+        return True
+
+    def serve(self):
+        """Starts gateway server."""
+
+        enable_auth = self.nvme_config.getboolean("config", "enable_auth")
+        gateway_addr = self.nvme_config.get("config", "gateway_addr")
+        gateway_port = self.nvme_config.get("config", "gateway_port")
+        grpc_max_workers = self.nvme_config.getint("config",
+                                                   "grpc_server_max_workers")
+
+        if enable_auth:
+            # Read in key and certificates for authentication
+            server_key = self.nvme_config.get("mtls", "server_key")
+            server_cert = self.nvme_config.get("mtls", "server_cert")
+            client_cert = self.nvme_config.get("mtls", "client_cert")
+            with open(server_key, "rb") as f:
+                private_key = f.read()
+            with open(server_cert, "rb") as f:
+                server_crt = f.read()
+            with open(client_cert, "rb") as f:
+                client_crt = f.read()
+
+            # Create appropriate server credentials
+            server_credentials = grpc.ssl_server_credentials(
+                private_key_certificate_chain_pairs=[(private_key, server_crt)],
+                root_certificates=client_crt,
+                require_client_auth=True,
+            )
+
+            # Create server and check for existing NVMeoF target configuration
+            self.server = grpc.server(
+                futures.ThreadPoolExecutor(max_workers=grpc_max_workers))
+            self.start_spdk()
+            self.persistent_config.restore(self.spdk_rpc, self.spdk_rpc_client)
+            pb2_grpc.add_NVMEGatewayServicer_to_server(self, self.server)
+
+            # Add secure port using crendentials
+            self.server.add_secure_port(
+                "{}:{}".format(gateway_addr, gateway_port), server_credentials)
+        else:
```

Comment:
nit: not sure we need this code duplication. Why couldn't the same code above just be moved out of the `if enable_auth` block?